File: framework/ui/src/io/intino/alexandria/ui/displays/components/BaseLocation.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.schemas.LocationCenter;
import io.intino.alexandria.schemas.LocationSetup;
import io.intino.alexandria.schemas.LocationZoomRange;
import io.intino.alexandria.ui.Asset;
import io.intino.alexandria.ui.displays.components.geo.PlaceMarkBuilder;
import io.intino.alexandria.ui.displays.notifiers.BaseLocationNotifier;
import io.intino.alexandria.ui.model.Geometry;

import java.net.URL;

public class BaseLocation<DN extends BaseLocationNotifier, B extends Box> extends AbstractBaseLocation<DN, B> {
    private URL icon = null;
    private Geometry value = null;

    public BaseLocation(B box) {
        super(box);
    }

    public Geometry value() {
        return value;
    }

    public BaseLocation<DN, B> value(String value) {
        return value(value != null ? Geometry.fromWkt(value) : null);
    }

    public BaseLocation<DN, B> value(Geometry location) {
        _value(location);
        refresh();
        return this;
    }

    public BaseLocation<DN, B> zoom(int zoom) {
        notifier.refreshZoom(zoom);
        return this;
    }

    public BaseLocation<DN, B> zoomRange(int min, int max) {
        notifier.refreshZoomRange(new LocationZoomRange().min(min).max(max));
        return this;
    }

    public BaseLocation<DN, B> center(double latitude, double longitude) {
        notifier.refreshCenter(new LocationCenter().lat(latitude).lng(longitude));
        return this;
    }

    @Override
    public void init() {
        super.init();
        setup();
        refresh();
    }

    @Override
    public void refresh() {
        notifier.refresh(value != null ? PlaceMarkBuilder.buildGeometry(value) : null);
    }

    protected <D extends BaseLocation> D _icon(URL icon) {
        this.icon = icon;
        return (D) this;
    }

    protected <D extends BaseLocation> D _value(Geometry value) {
        this.value = value;
        return (D) this;
    }

    protected <D extends BaseLocation> D _value(String value) {
        return _value(Geometry.fromWkt(value));
    }

    private void setup() {
        if (icon == null) return;
        LocationSetup setup = new LocationSetup();
        setup.icon(Asset.toResource(baseAssetUrl(), this.icon).toUrl().toString());
        notifier.setup(setup);
    }

}

File: framework/ui-elements/src/displays/components/BaseLocation.js
import React from "react";
import AbstractBaseLocation from "../../../gen/displays/components/AbstractBaseLocation";
import BaseLocationNotifier from "../../../gen/displays/notifiers/BaseLocationNotifier";
import BaseLocationRequester from "../../../gen/displays/requesters/BaseLocationRequester";
import { GoogleMap } from '@react-google-maps/api'
import PlaceMark from "./geo/PlaceMark";
import SearchDialog from "./geo/SearchDialog";
import GeoBehavior from "./behaviors/GeoBehavior";
import 'alexandria-ui-elements/res/styles/layout.css';

const styles = theme => ({
	map : {
		minHeight: "100%"
	}
});

export default class BaseLocation extends AbstractBaseLocation {

	constructor(props) {
		super(props);
		this.notifier = new BaseLocationNotifier(this);
		this.requester = new BaseLocationRequester(this);
		this.container = React.createRef();
		this.googleMapLayer = React.createRef();
        this.state = {
            ...this.state,
            location: null,
            icon: null,
            center: null,
            zoom: null,
        }
	};

	renderLayer = (content) => {
		const { classes } = this.props;

		window.setTimeout(() => this.resize(), 100);

		return (
			<div ref={this.container} className="layout flex" style={{position:'relative',height:"100%"}}>
			    {this.renderSearch()}
				<GoogleMap className={classes.map} zoom={GeoBehavior.zoom(this).defaultZoom}
						   center={GeoBehavior.center(this)} options={this.mapOptions()}
						   onLoad={this.registerMap.bind(this)}>
					<div ref={this.googleMapLayer} style={{width:'100%'}}/>
					{this.renderPlaceMark()}
					{content != null && content()}
				</GoogleMap>
			</div>
		);
	};

	renderSearch = () => {
	    if (!this.allowSearch()) return (<React.Fragment/>);
	    return (<SearchDialog map={this.getMap.bind(this)} classes={this.props.classes} mapOptions={this.mapOptions()}/>);
	};

	getMap = () => {
	    return this.map;
	};

	registerMap = (map) => {
	    this.map = map;
	};

	resize = function() {
		if (this.googleMapLayer.current == null) return;
		const container = this.container.current;
		const height = $(container).height();
		if (height == 0) {
		    window.setTimeout(() => this.resize(), 100);
		    return;
		}
		this.googleMapLayer.current.style.height = $(container).height() + "px";
	};

	renderPlaceMark = () => {
		const icon = this.state.icon;
		const location = this.state.location;
		if (location == null) return;
		return (
			<PlaceMark placeMark={{icon:icon,location:location}}/>
		);
	};

	setup = (info) => {
		this.setState({ icon: info.icon });
	};

	refresh = (location) => {
		this.setState({ location: location });
	};

    refreshCenter = (center) => {
        this.setState({ center });
    };

    refreshZoom = (value) => {
        const zoom = GeoBehavior.zoom(this);
        this.setState({ min: zoom.min, max: zoom.max, defaultZoom: value });
    };

    refreshZoomRange = (range) => {
        const zoom = GeoBehavior.zoom(this);
        this.setState({ min: range.min, max: range.max, defaultZoom: zoom.defaultZoom });
    };

    mapOptions = () => {
        const controls = this.props.controls != null ? this.props.controls : "";
        const all = controls.indexOf("all") != -1;
		return {
            zoomControl: controls.indexOf("zoom") != -1 || all,
            mapTypeControl: controls.indexOf("maptype") != -1 || all,
            scaleControl: controls.indexOf("scale") != -1 || all,
            streetViewControl: controls.indexOf("streetview") != -1 || all,
            rotateControl: controls.indexOf("rotate") != -1 || all,
            fullscreenControl: controls.indexOf("fullscreen") != -1 || all
        };
    };

    allowSearch = () => {
        const controls = this.props.controls != null ? this.props.controls : "";
        return controls.indexOf("all") != -1 || controls.indexOf("search") != -1;
    };
}